### PR TITLE
the first working version of transaction fetcher

### PR DESCRIPTION
**Pull request by konnov**
This PR presents a working implementation of transaction decoder. The core of this contribution is:

 - [x] `solarkraft/src/fetcher/callDecoder.ts` extracts Soroban contract calls from a Horizon API.
 - [x] `solarkraft/src/fetcher/storage.ts` is the basic storage data structure that is returned by `callDecoder`.
 - [x] `solarkraft/test/integration/callDecoder.test.ts` is an integration test that fetches data from the testnet and compares the results to the expected outputs.
 - [ ] `solarkraft/src/fetch.ts` is work-in-progress code, ignore it.

The integration test is fragile as the test data will disappear after 1 month or when the testnet is reset. We have to find a more stable solution. However, for the development purposes, it should work.

### CODE REVIEW

File: solarkraft/src/fetcher/callDecoder.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import { Maybe, just, none } from '@sweet-monads/maybe'
+import { OrderedMap } from 'immutable'
+
+/**
+ * Decode a contract call from a Horizon operation.
+ * @param op an operation
+ * @param matcher a quick matcher over the contractId to avoid expensive deserialization
+ */
+export async function extractContractCall(
+    op: any,
```
Comment (Kukovec) [inline solarkraft/src/fetcher/callDecoder.ts:23]: Is there no way to be more precise here?

File: solarkraft/src/fetcher/callDecoder.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        // is it the right way to extract the height?
+        const height = tx.ledger_attr
+
+        // The operation itself does not give us the ledger updates.
+        // Deserialize tx.result_meta_xdr.
+        const meta = sdk.xdr.TransactionMeta.fromXDR(
+            tx.result_meta_xdr,
+            'base64'
+        )
+        if (meta.switch() !== 3) {
```
Comment (Kukovec) [inline solarkraft/src/fetcher/callDecoder.ts:59]: Can you add a comment explaining what `3` represents here?

File: ContractExamples/contracts/setter/src/lib.rs [hunk 3]
```diff
@@ -64,7 +64,7 @@ impl SetterContract {
         env.storage().instance().set(&MY_BOOL, &v);
         log!(&env, "myBool: {}", v);
         // bump the lifetime
-        env.storage().instance().extend_ttl(50, 100);
+        env.storage().instance().extend_ttl(env.storage().max_ttl() - 100, env.storage().max_ttl());
```
Comment (thpani) [inline ContractExamples/contracts/setter/src/lib.rs:67]: Why do we subtract 100?

File: solarkraft/src/fetcher/storage.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     * The arguments passed to the method, in native JS.
+     */
+    methodArgs: any[]
+    /**
+     * The return value in native JS.
+     */
+    returnValue: any
+    /**
+     * Ordered mapping from field names to their native values (JS).
+     * This mapping contains values only for the fields that have been set
```
Comment (thpani) [inline solarkraft/src/fetcher/storage.ts:50]: So if I want all values after the tx, I have to merge `oldFields` and `fields`?

`updatedFields` may be a better name here.

File: solarkraft/src/fetcher/callDecoder.ts (hunk 1 above)
Comment (konnov) [reply solarkraft/src/fetcher/callDecoder.ts:23]: I am not actually sure. Stellar SDK has [Operation](https://stellar.github.io/js-stellar-sdk/js-stellar-base_src_operation.js.html), but it does not look like what Horizon API is returning in Stellar SDK. Some of the methods return `sdk.Horizon.ServerApi.OperationRecord` but it's not exactly compatible with what I need.

**Review by thpani — COMMENTED** [review]
Preliminary comments, I'm still looking at `callDecoder.ts`

File: ContractExamples/contracts/setter/src/lib.rs (hunk 3 above)
Comment (Kukovec) [reply ContractExamples/contracts/setter/src/lib.rs:67]: +1, I had a similar question, but just assumed I was OOTL

Comment (konnov) [reply ContractExamples/contracts/setter/src/lib.rs:67]: it's a threshold of when the TTL is to be updated. I am not exactly sure how it works, so it's a bit arbitrary.

File: solarkraft/src/fetcher/storage.ts (hunk 4 above)
Comment (konnov) [reply solarkraft/src/fetcher/storage.ts:50]: No, actually, `fields` contains all fields stored so far. It's basically the new state.

Comment (konnov) [reply solarkraft/src/fetcher/storage.ts:50]: Of course, if a contract removes a field from the storage, then `fields` will have fewer fields.

File: ContractExamples/contracts/setter/src/lib.rs (hunk 3 above)
Comment (thpani) [reply ContractExamples/contracts/setter/src/lib.rs:67]: Makes sense. Could you add a brief comment? It looks a bit arbitrary.

File: solarkraft/src/fetcher/storage.ts (hunk 4 above)
Comment (konnov) [reply solarkraft/src/fetcher/storage.ts:50]: I have updated the comments a bit

Comment (thpani) [reply solarkraft/src/fetcher/storage.ts:50]: Then I dont understand the comment (`contains values only for the fields that have been set by transactions`). Which fields aren't  set then?

File: solarkraft/src/fetcher/callDecoder.ts (hunk 2 above)
Comment (konnov) [reply solarkraft/src/fetcher/callDecoder.ts:60]: Done

File: solarkraft/src/fetcher/callDecoder.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+ * @param op an operation
+ * @param matcher a quick matcher over the contractId to avoid expensive deserialization
+ */
+export async function extractContractCall(
+    op: any,
+    matcher: (contractId: string) => boolean
+): Promise<Maybe<ContractCallEntry>> {
+    // https://developers.stellar.org/network/horizon/api-reference/resources/operations/object
+    if (op.function === 'HostFunctionTypeHostFunctionTypeInvokeContract') {
+        // This operation represents a smart contract call.
```
Comment (thpani) [inline solarkraft/src/fetcher/callDecoder.ts:28]: Can't you just return `none` if `!==`.
Bit hard to follow controlflow here with the indentation.

File: solarkraft/src/fetcher/callDecoder.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        if (!matcher(contractId)) {
+            // return quickly to avoid additional calls and deserialization
+            return none()
+        }
+
+        // continue with extracting transaction data and ledger updates
+        const method = params[1]
+        const methodArgs = params.slice(2)
+
+        const tx = await op.transaction()
```
Comment (thpani) [inline solarkraft/src/fetcher/callDecoder.ts:49]: This is weird. Is it really async?

File: ContractExamples/contracts/setter/src/lib.rs (hunk 3 above)
Comment (konnov) [reply ContractExamples/contracts/setter/src/lib.rs:70]: Added a comment. Actually, TTL is extended when it goes below the threshold, in this case, `max_ttl - 100`.

File: solarkraft/src/fetcher/callDecoder.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        // is it the right way to extract the height?
```
Comment (thpani) [inline solarkraft/src/fetcher/callDecoder.ts:50]: Is this a TODO?

File: solarkraft/src/fetcher/callDecoder.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        )
+        // extract transaction metadata for version 3
+        if (meta.switch() !== 3) {
+            // Is it possible to have older transaction metadata with Soroban?
+            // TODO: make it future-proof
+            console.error(
+                `Expected TransactionMetaV3, found ${meta.switch()}: txHash = ${txHash}`
+            )
+            return none()
+        } else {
```
Comment (thpani) [inline solarkraft/src/fetcher/callDecoder.ts:67]: Also here, I don't think you need to nest inside else, since you return above?

**Review by thpani — APPROVED** [review]
Remaining comments, rest lgtm

File: ContractExamples/contracts/setter/src/lib.rs (hunk 3 above)
Comment (konnov) [reply ContractExamples/contracts/setter/src/lib.rs:70]: Here is the fresh version. It's all modulo my understanding of how Soroban stores the storage in Stellar:

```js
    /**
     * Ordered mapping from field names to their native values (JS).
     * This mapping contains values only for the fields that have been created
     * or updated by a transaction in the past. It may happen that `fields` contains
     * fewer fields than `oldFields`, when the contract deletes some fields
     * from the storage. Also, `fields` may become empty when the storage
     * goes over TTL.
     */
```

File: solarkraft/src/fetcher/storage.ts (hunk 4 above)
Comment (konnov) [reply solarkraft/src/fetcher/storage.ts]: Here is the fresh version. It's all modulo my understanding of how Soroban stores the storage in Stellar:

```js
    /**
     * Ordered mapping from field names to their native values (JS).
     * This mapping contains values only for the fields that have been created
     * or updated by a transaction in the past. It may happen that `fields` contains
     * fewer fields than `oldFields`, when the contract deletes some fields
     * from the storage. Also, `fields` may become empty when the storage
     * goes over TTL.
     */
```

File: solarkraft/src/fetcher/callDecoder.ts (hunk 6 above)
Comment (konnov) [reply solarkraft/src/fetcher/callDecoder.ts:49]: yeah, it looks async. Basically, SDK makes another call to the server.

File: solarkraft/src/fetcher/callDecoder.ts (hunk 7 above)
Comment (konnov) [reply solarkraft/src/fetcher/callDecoder.ts:50]: I mean, yeah. I hope someone more knowledgeable would read it tells me to fix it :)

Comment (konnov) [reply solarkraft/src/fetcher/callDecoder.ts]: added `TODO:` for the future

File: solarkraft/src/fetcher/callDecoder.ts (hunk 8 above)
Comment (konnov) [reply solarkraft/src/fetcher/callDecoder.ts]: yeah, good catch. It grew too large organically. Pruned it

File: solarkraft/src/fetcher/storage.ts (hunk 4 above)
Comment (thpani) [reply solarkraft/src/fetcher/storage.ts]: Oh, you meant values that ever have been set! Understood, lg

**Comment by konnov** [comment]
I will merge this one to continue with the storage.
